### PR TITLE
Add Triggers documentation

### PR DESCRIPTION
**Pull request by appetrosyan**
Signed-off-by: Aleksandr <a-p-petrosyan@yandex.ru>Signed-off-by: Aleksandr <a-p-petrosyan@yandex.ru>

### CODE REVIEW

File: src/guide/advanced/triggers.md [hunk 1]
```diff
@@ -0,0 +1,244 @@
...
+
+let filter = …
+
+```
+
+the event filter is where we need to spend some time and think. So far we've seen the `Pipeline` variety of filters. This time around, the filter is a `Data` kind, which is itself a tuple with a single variant, which is a `FilterOpt` of an `EntityFilter`.
+
+`FilterOpt` stands for Optional Filter. It can either `AcceptAll`, or accept `BySome` of another `Filter`. An `EntityFilter` is a filter that matches `ByAccount` in our case, but can match by many other means. It wraps an `AccountFilter`, which matches on various events produced on accounts. We need to listen for an event that occurs when Alice's asset is reduced: so we need `AccountEventFilter::ByAsset`. That itself is an `AssetEventFilter`. It should filter `ByRemoved` (but not `ByDeleted`). **It should be noted that the names are subject to change, as they are confusing.**
+
+This gets us nowhere, and a big problem in our current API, that we're trying to solve. Instead we should work bottom up.
```
Comment (WRRicht3r) [inline src/guide/advanced/triggers.md:165]:
```suggestion
This gets us nowhere and is a big problem in our current API, that we're trying to solve. Instead, we should work bottom-up.
```

File: src/guide/advanced/triggers.md [hunk 2]
```diff
@@ -0,0 +1,244 @@
...
+
+After this somewhat laborious filter combination, we can finally create an `Action`
+
+```rust
+let action = Action {
+    executable, repeats, technical_account, filter
+}
+```
+
+and finally create an instance of a `Trigger`.
```
Comment (WRRicht3r) [inline src/guide/advanced/triggers.md:199]:
```suggestion
and then create an instance of a `Trigger`.
```

File: src/guide/advanced/triggers.md [hunk 3]
```diff
@@ -0,0 +1,244 @@
...
+
+```rust
+let trigger = Trigger {
+    id, action, metadata
+}
+```
+
+Finally in order to get said trigger onto the blockchain, we create a transaction with the following single instruction.
```
Comment (WRRicht3r) [inline src/guide/advanced/triggers.md:207]:
```suggestion
Finally, in order to get said trigger onto the blockchain, we create a transaction with the following single instruction.
```

File: src/guide/advanced/triggers.md [hunk 4]
```diff
@@ -0,0 +1,244 @@
...
+
+The technical details can be summarised in the following. First, the normal instructions that either got submitted from WASM or directly get executed. If there were any triggers that should have been registered, they get registered now. Then using the total set of events that got generated during the execution of instructions, the triggers including some that got registered just this round get executed. Not in the next block, but this block. Then the events from the execution of the triggers are recorded in the _next_ block.
+
+::: info
+
+The reason why the events caused by the execution of triggers get processed in the next block is simple. We don't want two triggers to inadvertently cause an infinite loop of instruction execution, and break consensus.
+
+:::
+
+Now, each time alice drinks some tea, the Mad Hatter will pour in a whole new cup. The keen eyed among you will see that the amount that alice drinks is irrelevant to how much tea will be transferred. So Alice may well take a sip and be poured a whole new cup's worth.
```
Comment (WRRicht3r) [inline src/guide/advanced/triggers.md:224]:
```suggestion
Now, each time Alice drinks some tea, the Mad Hatter will pour in a whole new cup. The keen-eyed among you will see that the amount that Alice drinks is irrelevant to how much tea will be transferred. So Alice may well take a sip and be poured a whole new cup's worth.
```

**Review by WRRicht3r — CHANGES_REQUESTED** [review]

**Comment by 0x009922** [comment]
Why is it located in `advanced` dir? Why not in the appendix one?

Anyway, you should add an appropriate record into the sidebar links list. It is here:

https://github.com/hyperledger/iroha-2-docs/blob/f074c7aa0f3f40bd7a8b77d6fb7f3bcf705b8869/.vitepress/config.ts#L30-L101

File: src/guide/advanced/triggers.md [hunk 5]
```diff
@@ -0,0 +1,244 @@
...
+```rust
+enum Repeats {
+  Indefinitely,
+  Exactly(u32),
+}
+```
+
+A technical account is the account that would (in theory) be responsible for the execution environment and be the authority for `Instruction` execution. You can (for now) leave this to be the account that registered the trigger, that is, if you followed the tutorial, `alice@wonderland`. However, later on we will show you why you'd want to create a brand new account for those purposes.
+
+A filter is what determines the _kind_ of trigger that you're dealing with. All triggers respond to events, but differently, and that difference  comes from the `EventFilter`. The reason for this architectural choice is simple. People use `EventFilters` in their front-end code all the time. So you don't need to learn how to do two different things, you just use the same knowledge in two different places.
```
[the remark this answers is not in the document]
Comment (appetrosyan) [reply src/guide/advanced/triggers.md:58]: Rephrased.

File: src/guide/advanced/triggers.md [hunk 6]
```diff
@@ -0,0 +1,244 @@
...
+2. Timed triggers
+3. By-call triggers
+4. Block-based triggers
+
+Each of these shall be covered in its own right, and some information shall be provided on how to use each of them in detail. As this tutorial is under construction the triggers that have not yet been implemented will be described only theoretically. Let's go over the basics of each.
+
+
+### Event triggers
+
+All triggers are event triggers in a sense. But the events in question are the ones that you'd get once something interesting happened in the blockchain: an asset got minted, an account got registered, etc. For now, only un-scoped global triggers are supported. That means, your trigger can see every account on the blockchain. This is both good and bad.
```
[the remark this answers is not in the document]
Comment (appetrosyan) [reply src/guide/advanced/triggers.md:73]: Reworded.

File: src/guide/advanced/triggers.md [hunk 7]
```diff
@@ -0,0 +1,244 @@
...
+
+It's good in that you don't need to worry about registering your trigger in the proper name-space (for now), but it also means that your trigger processes almost every event in the blockchain.
```
[the remark this answers is not in the document]
Comment (appetrosyan) [reply src/guide/advanced/triggers.md:75]: Removed.

File: src/guide/advanced/triggers.md (hunk 4 above)
Comment (appetrosyan) [reply src/guide/advanced/triggers.md:224]: Reworded

**Pull request merged**